src/utilities/loading-utils.ts

```diff
@@ -17,21 +30,97 @@ export function getCurrentLoadingMessage() {
 }
 
 export function showLoadingMessage(msg: string) {
+  const timeStamp = getTimestamp();
+  if(timeStamp === null) {
+    console.error(`Timestamp for message "${msg}" is null.`);
+    return;
+  }
   if (DEBUG_LOADING) {
-    // eslint-disable-next-line no-console
-    console.log(`Loading @${getTimestamp()}ms: ${msg}`);
+    console.log(`Loading @${timeStamp}ms: ${msg}`);
   }
   const existingMessages = sessionStorage.getItem(sessionStorageMessageItem);
   sessionStorage.setItem('loading-message', existingMessages + msg + messageSeparator);
+  const msgStart = msg + " Start";
+  loadingMeasurements[msgStart] = timeStamp; //Start time
 }
 
 export function removeLoadingMessage(msg: string) {
+  const timeStamp = getTimestamp();
+  if (timeStamp === null) {
+    console.error(`Timestamp for message "${msg}" is null.`);
+    return;
+  }
   if (DEBUG_LOADING) {
-    // eslint-disable-next-line no-console
-    console.log(`Loading @${getTimestamp()}ms: Done with: ${msg}`);
+    console.log(`Loading @${timeStamp}ms: Done with: ${msg}`);
   }
   const existingMessages = sessionStorage.getItem(sessionStorageMessageItem);
   if (existingMessages) {
     sessionStorage.setItem('loading-message', existingMessages.replace(msg + messageSeparator, ''));
   }
+  const msgEnd = msg + " End";
+  loadingMeasurements[msgEnd] = timeStamp; //End time
+}
+
+export function getLoadingMeasurements(){
+  return { loadingMeasurements };
+}
+
+
+export function logLoadingAndDocumentMeasurements(documents: DocumentsModelType,
+  curriculumDocSections: SectionModelType[],
+  primaryDocument?: DocumentModelType ){
+  const startTime = performance.now();
+
+  const totalNumDocumentsLoaded = documents.all.length;
+  const totalNumTilesLoaded = documents.all.reduce((total: number, doc: DocumentModelType) => {
+    return total + (doc.content?.tileMap.size || 0);
+  }, 0);
+
+  const primaryDocTilesByType = primaryDocument?.content?.getAllTilesByType() as any;
+  const primaryDocNumTilesByType = countTileKeys(primaryDocTilesByType);
```

Comment:
```suggestion
  const primaryDocNumTilesByType = primaryDocTilesByType ?? countTileKeys(primaryDocTilesByType);
```
This is safer since primaryDocTilesByType might be undefined based on line 79. This change might also let you remove the `as any` case on line 79.

Reply:
The `as any` still remains, but I implemented that change.